zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncCallbacks.java

```diff
@@ -170,7 +170,17 @@ public void callback(int rc, String path, Object ctx) {
       }
 
       if (ctx != null && ctx instanceof ZkAsyncCallMonitorContext) {
-        ((ZkAsyncCallMonitorContext) ctx).recordAccess(path);
+        ZkAsyncCallMonitorContext monitor = (ZkAsyncCallMonitorContext) ctx;
+        monitor.recordAccess(path);
+        // Record failure if the return code isn't 0 (i.e., OK), and this object
+        // is one of the five callback types derived from DefaultCallback
```

Comment:
This check looks verbose. How about defining a reportFailure boolean in the DefaultCallback class. By default it is true. Then you overwrite it for SyncCallbackHandler (it seems to be the only one that you don't want to record failure, Why is that BTW?)
Then you can read this boolean field here and execute accordingly.

Reply:
Thanks, I will use a boolean variable to simplify the check. And thanks for catching SyncCallbackHandler, not recording it wasn't intentional but since there wasn't an equivalent recording in ZkClient I missed it. I will add this to the check too. Only one question, for GetDataCallbackHandler and ExistsCallbackHandler I don't record a failure if the return code is NONODE (given the logic in their sync equivalent methods in ZkClient), should I again make this exception for SyncCallbackHandler? Or this doesn't make sense because SyncCallbackHandler checks the root path?

Reply:
Please refer to the ZK document to see if Sync ever throws NO NODE error. I believe it does not. But I might be wrong : )

One more thing, given this logic, I think you can make the check more flexible instead of just a boolean field. Something like this,

boolean shouldRecordFailure(int rc) {
  // default return true if rc != 0;
  // special handle get/exists calls in the child class implementation.
}

Not a must. If it becomes too complicated, then you shall make an alternative design. My intention is mainly to reduce duplicate code when we have a clean OO design.

Reply:
This is very helpful, thank you. I'll use it in the next commit.